### PR TITLE
convert basic statistics pp to use the new data object 

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
Currently, four tests of basic statistics are failed. 
```
framework/PostProcessors/BasicStatistics.timeDepOutCsv FAILED (CWD /Users/wangc/projects/wangc/raven METHOD ? Expected files not created /Users/wangc/projects/wangc/raven/tests/framework/PostProcessors/BasicStatistics/timeDepOutCsv/fileOut.csv /Users/wangc/projects/wangc/raven/tests/framework/PostProcessors/BasicStatistics/timeDepOutCsv/1-PlotStatsT2_scatter-scatter-scatter.png)
framework/PostProcessors/BasicStatistics.timeDependentAsyncHistories FAILED (CWD /Users/wangc/projects/wangc/raven METHOD ? Expected files not created /Users/wangc/projects/wangc/raven/tests/framework/PostProcessors/BasicStatistics/basicStatisticsTimeDependentAsynchronousHistories/output_basicStatistics_timedep_PP1.csv)
framework/PostProcessors/BasicStatistics.variationCoefficient FAILED (
Could not find match for row "2" in Gold:
vc_x0         inf
vc_y0    0.011313
vc_z0    0.150734
vc_x     0.176795
vc_y     0.011246
vc_z     0.156545
dtype: float64)
framework/PostProcessors/BasicStatistics.general FAILED (
Could not find match for row "2" in Gold:
skew_time                      NaN
skew_x0              -7.411787e-01
```

The first one ```BasicStatistics.timeDepOutCsv``` is related to IOStep,
The second one is related to postprocessor ```InterfacedPostProcessor|HistorySetSync```
The third and forth are related to the csv diff, which can not handle __NAN__ and __inf__.

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [ ] 1. Review all computer code.
- [ ] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [ ] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [ ] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [ ] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [ ] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [ ] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [ ] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: framework/Models/ExternalModel.py [hunk 1]
```diff
@@ -252,9 +252,14 @@ def evaluateSample(self, myInput, samplerType, kwargs):
...
     rlz = {}
-    rlz.update(inRun)
+    # FIXME: The results may contain the variables from input space, and
+    # the external model may change the values, we should raise
+    # an warning, and have a similar method as the code does
     rlz.update(result)
+    rlz.update(inRun)
     rlz.update(kwargs)
+    for k,v in rlz.items():
+      rlz[k] = np.atleast_1d(v)
```
Comment (PaulTalbot-INL) [inline framework/Models/ExternalModel.py]: A better version of this is pending in the AdaptiveSobol, as well as the Optimizer merges.

File: framework/PostProcessors/BasicStatistics.py [hunk 2]
```diff
@@ -174,50 +172,29 @@ def inputToInternal(self, currentInp):
...
-          inputDict['targets'][targetP] = currentInput.getParam('output', targetP, nodeId = 'ending')
-        else:
-          self.raiseAnError(IOError, self, 'Target ' + targetP + ' has not been found in data object '+currentInput.name)
-    else:
+
+    metadata =  currentInput.getMeta(pointwise=True)
+    inputList = []
+    if currentInput.type == 'PointSet':
+      inputDict = {}
+      inputDict['targets'] = currentInput.getVarValues(self.parameters['targets'])
```
Comment (PaulTalbot-INL) [inline framework/PostProcessors/BasicStatistics.py]: We should improve this later to not use getVarValues, but for now it is okay.  Ideally, we can operate on the data directly without transforming it into dicts first.

File: framework/PostProcessors/BasicStatistics.py [hunk 3]
```diff
@@ -174,50 +172,29 @@ def inputToInternal(self, currentInp):
...
-          inputDict['timeDepData'][pivotParameter[ts]]['targets'][targetP] = outputValues[:,ts]
-          if cnt == 0:
-            inputDict['timeDepData'][pivotParameter[ts]]['metadata'] = metadata
+      self.dynamic = True
+      self.pivotValue = currentInput.asDataset()[self.pivotParameter].values
+      slices = currentInput.sliceByIndex(self.pivotParameter)
+      for sliceData in slices:
+        inputDict = {}
+        inputDict['metadata'] = metadata
+        inputDict['targets'] = dict((target, sliceData[target]) for target in self.parameters['targets'])
```
Comment (PaulTalbot-INL) [inline framework/PostProcessors/BasicStatistics.py]: There may be "NaN" in sliceData[target] if the histories are not aligned.  Calling "dropna" can fix this.  However, we require alignment, right?  Did "checkAlignement" get called before this?

File: framework/PostProcessors/BasicStatistics.py [hunk 4]
```diff
@@ -270,7 +241,12 @@ def _handleInput(self, paramInput):
     for child in paramInput.subparts:
       tag = child.getName()
       #because percentile is strange (has an attached parameter), we address it first
+      if tag in ['percentile'] + self.scalarVals + self.vectorVals:
+        if 'prefix' not in child.parameterValues:
+          self.raiseAnError(IOError, "Not prefix is provided for node: ", tag)
```
Comment (PaulTalbot-INL) [inline framework/PostProcessors/BasicStatistics.py]: "Not" should be "No"

File: framework/PostProcessors/BasicStatistics.py [hunk 5]
```diff
@@ -279,80 +255,36 @@ def _handleInput(self, paramInput):
...
-            self.toDo['percentile'][reqPercent].update(targets)
-          else:
-            self.toDo['percentile'][reqPercent] = set(targets)
+          strPercent = [child.parameterValues['percent']]
+        for reqPercent in strPercent:
+          self.toDo[tag].append({'targets':set(targets),
+                                 'prefix':prefix,
+                                 'percent':reqPercent})
+          #TODO: remove
+          #self.parameters['percentile_map'][floatPercentile[0]] = '5'
```
Comment (PaulTalbot-INL) [inline framework/PostProcessors/BasicStatistics.py]: Did you want to remove this?

File: scripts/conversionScripts/toBasicStatsPP.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import xml.etree.ElementTree as ET
+import xml.dom.minidom as pxml
+import os
+
+def convert(tree,fileName=None):
+  """
+    Converts input files to be compatible with merge request ###, where BasicStatistics is given the power
+    to be more nodalized than before.
```
Comment (PaulTalbot-INL) [inline scripts/conversionScripts/toBasicStatsPP.py]: This description should be updated.

File: tests/framework/PostProcessors/BasicStatistics/factorial_analytic.xml [hunk 7]
```diff
@@ -143,6 +171,96 @@
...
+                 cov_z_z,
+                 vsen_x_x,
+                 vsen_x_y,
+                 vsen_x_z,
+                 vsen_y_x,
+                 vsen_y_y,
+                 vsen_y_z,
+                 vsen_z_x,
+                 vsen_z_y,
+                 vsen_z_z</Group>
```
Comment (PaulTalbot-INL) [inline tests/framework/PostProcessors/BasicStatistics/factorial_analytic.xml]: maybe it's not worth changing, but we should create these groups without the newlines, as it clogs up the input file.

File: framework/PostProcessors/BasicStatistics.py (hunk 2 above)
Comment (alfoa) [reply framework/PostProcessors/BasicStatistics.py]: agree

File: framework/PostProcessors/BasicStatistics.py [hunk 8]
```diff
@@ -230,21 +207,15 @@ def initialize(self, runInfo, inputs, initDict):
...
-        for entry in self.toDo[vector]:
+    for metricName in self.scalarVals + self.vectorVals:
+      if metricName in self.toDo.keys():
+        for entry in self.toDo[metricName]:
           self.allUsedParams.update(entry['targets'])
-          self.allUsedParams.update(entry['features'])
+          try:
+            self.allUsedParams.update(entry['features'])
+          except KeyError:
+            pass
```
Comment (alfoa) [inline framework/PostProcessors/BasicStatistics.py]: Con you explain this a bit better?

File: framework/PostProcessors/BasicStatistics.py [hunk 9]
```diff
@@ -270,7 +241,12 @@ def _handleInput(self, paramInput):
     for child in paramInput.subparts:
       tag = child.getName()
       #because percentile is strange (has an attached parameter), we address it first
+      if tag in ['percentile'] + self.scalarVals + self.vectorVals:
+        if 'prefix' not in child.parameterValues:
+          self.raiseAnError(IOError, "Not prefix is provided for node: ", tag)
+        prefix = child.parameterValues['prefix']
       if tag == 'percentile':
+        #get the prefix
```
Comment (alfoa) [inline framework/PostProcessors/BasicStatistics.py]: this comment should be probably moved up

File: framework/PostProcessors/BasicStatistics.py [hunk 10]
```diff
@@ -279,80 +255,36 @@ def _handleInput(self, paramInput):
...
-            self.toDo[vector] = []
-          self.toDo[vector].append({'targets':set(tnode.value),
-                                 'features':set(fnode.value)})
+        # we're storing toDo[tag] as a list of dictionaries.  This is because the user might specify multiple
+        #   nodes with the same metric (tag), but with different targets and features.  For instance, the user might
+        #   want the sensitivity of A and B to X and Y, and the sensitivity of C to W and Z, but not the sensitivity
+        #   of A to W.  If we didn't keep them separate, we could potentially waste a fair number of calculations.
+        self.toDo[tag].append({'targets':set(tnode.value),
+                               'features':set(fnode.value),
+                               'prefix':prefix})
```
Comment (alfoa) [inline framework/PostProcessors/BasicStatistics.py]: good catch

File: framework/PostProcessors/BasicStatistics.py [hunk 11]
```diff
@@ -1045,91 +800,80 @@ def getCovarianceSubset(desired):
...
+        if metric in self.scalarVals:
+          for targetP in targetDict['targets']:
+            varName = prefix + '_' + targetP
+            outputDict[varName] = np.atleast_1d(calculations[metric][targetP])
+        #TODO someday we might need to expand the "skipped" check to include scalars, but for now
+        #   the only reason to skip is if an invalid matrix is requested
+        #if matrix block, extract desired values
+        else:
+          #check if it was skipped for some reason
+          skip = self.skipped.get(metric, None)
```
Comment (alfoa) [inline framework/PostProcessors/BasicStatistics.py]: Out of curiosity...when and why a metric can be skipped?

File: framework/PostProcessors/BasicStatistics.py [hunk 12]
```diff
@@ -1045,91 +800,80 @@ def getCovarianceSubset(desired):
...
-        elif metric in ['sensitivity','NormalizedSensitivity','VarianceDependentSensitivity']:
-          outputDict[metric] = {}
-          for entry in params:
-            #check if it was skipped for some reason
-            if entry == self.skipped.get(metric,None):
-              self.raiseADebug('Metric',metric,'was skipped for parameters',entry,'!  See warnings for details.  Ignoring...')
-              continue
-            for target in entry['targets']:
-              outputDict[metric][target] = dict((feature,calculations[metric][target][feature]) for feature in entry['features'])
+                outputDict[varName] = np.atleast_1d(calculations[metric]['matrix'][targetIndex,featureIndex])
```
Comment (alfoa) [inline framework/PostProcessors/BasicStatistics.py]: the multiple entries for a metric is available for jjust some relational metrics and not for others?

File: framework/PostProcessors/BasicStatistics.py [hunk 13]
```diff
@@ -1045,91 +800,80 @@ def getCovarianceSubset(desired):
...
-          for feature,value in wrt.items():
-            self.raiseADebug('     ',feature+':',value)
-      else:
-        self.raiseADebug('   ',valueDict)
-
   def run(self, inputIn):
     """
       This method executes the postprocessor action. In this case, it computes all the requested statistical FOMs
-      @ In,  inputIn, object, object contained the data to process. (inputToInternal output)
+      @ In,  inputIn, object, object contained the data to process. (enputToInternal output)
```
Comment (alfoa) [inline framework/PostProcessors/BasicStatistics.py]: enput

File: framework/PostProcessors/BasicStatistics.py [hunk 14]
```diff
@@ -1045,91 +800,80 @@ def getCovarianceSubset(desired):
...
-      outputDict = OrderedDict()
       self.raiseADebug('BasicStatistics Pivot-Dependent output:')
-      for pivotParamValue in inputAdapted['timeDepData'].keys():
-        self.raiseADebug('Pivot Parameter Value: ' + str(pivotParamValue))
-        outputDict[pivotParamValue] = self.__runLocal(inputAdapted['timeDepData'][pivotParamValue])
-
+      outputList = []
+      for inputDict in inputAdapted:
+        outputList.append(self.__runLocal(inputDict))
+      outputDict = dict((var,list()) for var in outputList[0].keys())
```
Comment (alfoa) [inline framework/PostProcessors/BasicStatistics.py]: you might use a defaultdict comprehension here (maybe in the future)

**Comment by alfoa** [comment]
I checked the regolding...As far as I can say, they look fine... => approved

File: framework/Models/ExternalModel.py (hunk 1 above)
Comment (wangcj05) [reply framework/Models/ExternalModel.py]: fixed.

File: framework/PostProcessors/BasicStatistics.py (hunk 2 above)
Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: I agree, and I put a __FIXME__ for the future development.

File: framework/PostProcessors/BasicStatistics.py (hunk 3 above)
Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: fixed.

File: framework/PostProcessors/BasicStatistics.py (hunk 4 above)
Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: fixed

File: framework/PostProcessors/BasicStatistics.py (hunk 5 above)
Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: removed.

File: scripts/conversionScripts/toBasicStatsPP.py (hunk 6 above)
Comment (wangcj05) [reply scripts/conversionScripts/toBasicStatsPP.py]: updated.

File: tests/framework/PostProcessors/BasicStatistics/factorial_analytic.xml (hunk 7 above)
Comment (wangcj05) [reply tests/framework/PostProcessors/BasicStatistics/factorial_analytic.xml]: Not sure, it is easy for me to check the variables.

File: framework/PostProcessors/BasicStatistics.py (hunk 8 above)
Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: Yes, for scalar metric, we store the variable as targets only, for matrix metric, we store the variables as targets and features, respectively.

File: framework/PostProcessors/BasicStatistics.py (hunk 9 above)
Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: fixed.

File: framework/PostProcessors/BasicStatistics.py (hunk 11 above)
Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: for example, not variables,

File: framework/PostProcessors/BasicStatistics.py (hunk 12 above)
Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: Sorry, I did not get your point.

File: framework/PostProcessors/BasicStatistics.py (hunk 13 above)
Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: fixed.

File: framework/PostProcessors/BasicStatistics.py (hunk 3 above)
Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: I added, but the test time_dep.xml failed, which should not fail based on my understanding. Could you take a look at that @PaulTalbot-INL

**Comment by wangcj05** [comment]
Comments are fixed, but the test time_dep.xml failed, because of the check alignment. Maybe I did not use it correctly.

File: framework/PostProcessors/BasicStatistics.py (hunk 3 above)
Comment (PaulTalbot-INL) [reply framework/PostProcessors/BasicStatistics.py]: It's hard for me to take a look at when I don't know where or how you used it, or what the error was.  Can you provide some additional reporting?  The method was made specifically for your application, so it needs to work for it.

**Comment by wangcj05** [comment]
We have fixed the checkIndexAlignment. @alfoa Do you have any comments?

**Comment by wangcj05** [comment]
test time_dep.xml is passing now.

File: framework/PostProcessors/BasicStatistics.py (hunk 3 above)
Comment (PaulTalbot-INL) [reply framework/PostProcessors/BasicStatistics.py]: Fixed.  Good work.

File: framework/PostProcessors/BasicStatistics.py (hunk 11 above)
Comment (alfoa) [reply framework/PostProcessors/BasicStatistics.py]: I honestly do not get it. Not enough variables? You mean dependent FOMs? (e.g. covariance for the VarianceBasedSensitivity matrix?)?

Comment (wangcj05) [reply framework/PostProcessors/BasicStatistics.py]: Based on the code, for the matrix we compute, if there is missing targets or features, the matrix computation will be skipped.

**Comment by wangcj05** [comment]
@PaulTalbot-INL can you do the merge?

**Comment by PaulTalbot-INL** [comment]
Done.

**Pull request merged**